File: apps/sync/lib/chalk_sync/whiteboard_v1/reducer.ex
defmodule ChalkSync.WhiteboardV1.Reducer do
  @moduledoc "Pure Excalidraw 0.18.1 element merge used by whiteboard-v1."

  @spec merge([map()], [map()]) :: [map()]
  def merge(current, incoming) when is_list(current) and is_list(incoming) do
    current
    |> Map.new(&{&1["id"], &1})
    |> merge_incoming(incoming)
    |> Map.values()
    |> Enum.sort_by(&{&1["index"], &1["id"]})
  end

  defp merge_incoming(elements, incoming) do
    Enum.reduce(incoming, elements, &merge_candidate/2)
  end

  defp merge_candidate(candidate, merged) do
    existing = Map.get(merged, candidate["id"])

    if is_nil(existing) or wins?(candidate, existing),
      do: Map.put(merged, candidate["id"], candidate),
      else: merged
  end

  defp wins?(candidate, existing) do
    if candidate["version"] == existing["version"],
      do: candidate["version_nonce"] < existing["version_nonce"],
      else: candidate["version"] > existing["version"]
  end
end

File: packages/whiteboard/src/collab/reducer.ts
import type { WhiteboardWireElement } from "./wire.js";

export function mergeWhiteboardElements(current: readonly WhiteboardWireElement[], incoming: readonly WhiteboardWireElement[]): readonly WhiteboardWireElement[] {
  const merged = new Map(current.map((element) => [element.id, element]));

  for (const candidate of incoming) {
    const existing = merged.get(candidate.id);
    if (!existing || wins(candidate, existing)) {
      merged.set(candidate.id, candidate);
    }
  }

  return [...merged.values()].sort(compareElementOrder);
}

function wins(candidate: WhiteboardWireElement, existing: WhiteboardWireElement): boolean {
  if (candidate.version !== existing.version) return candidate.version > existing.version;
  return candidate.version_nonce < existing.version_nonce;
}

function compareElementOrder(left: WhiteboardWireElement, right: WhiteboardWireElement): number {
  const byIndex = left.index.localeCompare(right.index);
  return byIndex === 0 ? left.id.localeCompare(right.id) : byIndex;
}
